cellpylib/ca_functions.py

```diff
@@ -84,13 +84,21 @@ def binary_rule(neighbourhood, rule, scheme=None):
     the NKS convention). If 'nks' is provided for the scheme parameter, the NKS convention is used for listing the 
     neighbourhoods.
     :param neighbourhood: a binary array of length 2r + 1
-    :param rule: an int indicating the cellular automaton rule number
+    :param rule: an int or a binary array indicating the cellular automaton rule number
     :param scheme: can be None (default) or 'nks'; if 'nks' is given, the rule numbering scheme used in NKS is used
+    :param powers_of_two: a pre-computed array containing the powers of two; can be None (default) or an array of length
+     len(neighbourhood); if an array is given, it will used to speed up the calculation of state_int
     :return: the result, 0 or 1, of applying the given rule on the given state
     """
-    state_int = bits_to_int(neighbourhood)
-    n = 2**len(neighbourhood)
-    rule_bin_array = int_to_bits(rule, n)
+    if powers_of_two is None:
+        state_int = bits_to_int(neighbourhood)
+    else:
+        state_int = neighbourhood.dot(powers_of_two)
+    n = 2 ** len(neighbourhood)
+    if isinstance(rule, np.ndarray):
+        rule_bin_array = rule
```

Comment:
The `rule_bin_array` needs to be zero padded so that it has a minimum length, as is done in the `int_to_bits` function. For example:
```
radius = 1
rule = np.array([1, 1, 1, 1, 0])
rule_bin_array = np.pad(rule, ((2 ** (radius * 2 + 1)) - len(rule), 0), 'constant')
```
The code later on in the function expects the `rule_bin_array` to be of length 2^n, where n is the neighbourhood size (i.e. 2r+1). It slows things down to do this padding in the function, so it's best if it is done up front. This means that there should be a check that ensures the rule, when it's an array, has the right length. Alternatively, we can zero pad for the user if required.